### PR TITLE
Reduce the prevalence of `None` as a return value

### PR DESCRIPTION
**Pull request by kurtmckee**
_Supersedes #152_

Changes
-------

-   ``AuthState.introspect_token()`` will no longer return ``None`` if the token is not active.

    Instead, a new exception, ``InactiveTokenError``, will be raised.
    ``InactiveTokenError`` is a subclass of ``ValueError``.

    Existing code that calls ``AuthState.introspect_token()`` no longer returns ``None``, either,
    but will instead raise ``ValueError``, a subclass of ``ValueError``,
    or a ``globus_sdk.GlobusAPIError``:

    *   ``AuthState.get_authorizer_for_scope``
    *   ``AuthState.effective_identity``
    *   ``AuthState.identities``

### CODE REVIEW

File: src/globus_action_provider_tools/authentication.py
Comment (sirosen) [inline src/globus_action_provider_tools/authentication.py:125]:
```suggestion
            raise InactiveTokenError("The token is invalid.")
```

Small personal preference here.
Possible reasons for being invalid that I know of:
- the string is not an Auth-issued token string -- maybe it doesn't even parse internally, etc.
- the token expired
- the token was revoked explicitly by the user
- the token was not explicitly revoked, but our team found it in public and revoked it (e.g. we find a refresh token in someone's GitHub repo)

So I'd just say "it's invalid" and leave the possibility of revocation as something covered by that statement.

File: src/globus_action_provider_tools/authentication.py
Comment (sirosen) [inline src/globus_action_provider_tools/authentication.py:255]: I'm reading the comment block above and feel like we can improve this message.
It feels like it's telling us "there was a cache miss", but the more complete message is
"there was a cache miss, so we _did_ do the callout, and it failed to get matching data"

I'm going to suggest we simply label these as the "cold" and "hot" cache paths, and trust that it provides enough information for us to find our way back to this context if necessary. 

So another suggestion follows this, but first:

```suggestion
                raise ValueError("Dependent tokens did not match request (cold cache)")
```

File: src/globus_action_provider_tools/authentication.py
Comment (sirosen) [inline src/globus_action_provider_tools/authentication.py:267]:
```suggestion
                raise ValueError("Dependent tokens did not match request (hot cache)")
```

**Review by sirosen — CHANGES_REQUESTED** [review]
Some minor tweaks to apply, but I want to hold off on approving and merging until after we do a release.

Although we're now fairly certain that this won't cause impactful changes for known AP Tools users, I think including it in the next release would not match some of the messages I've sent to other teams about what to expect out of our next two releases.

File: changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst
Comment (kurtmckee) [reply changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst:10]: `InactiveTokenError` and `InvalidTokenScopesError` are subclasses of `ValueError`...?

File: src/globus_action_provider_tools/authentication.py
Comment (kurtmckee) [reply src/globus_action_provider_tools/authentication.py:125]: Done.

File: src/globus_action_provider_tools/authentication.py
Comment (kurtmckee) [reply src/globus_action_provider_tools/authentication.py:267]: Done.

File: src/globus_action_provider_tools/authentication.py
Comment (kurtmckee) [reply src/globus_action_provider_tools/authentication.py:255]: Done.

Comment (kurtmckee) [reply src/globus_action_provider_tools/authentication.py:255]: Actually, I'm unresolving this. I don't think the terms "cold" and "hot" cache are helpful here, because in all failure cases we've submitted an API request to Auth.

I think these can both just say "Dependent tokens did not match request". The rest is simply just cache-busting details.

File: changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst
Comment (sirosen) [reply changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst:10]: Right, but the text says that `ValueError` is a subclass of `ValueError`, unless I'm very confused. (And if I am, I'll count that as a request for a rephrase!)

File: src/globus_action_provider_tools/authentication.py
Comment (sirosen) [reply src/globus_action_provider_tools/authentication.py]: Yeah, I'm down for that. No distinction between the two errors! :+1:

File: changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst
Comment (sirosen) [reply changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst:10]: Oh! I see now. It's a list of options, of length 3,
- `ValueError`
- a subclass of `ValueError`
- `globus_sdk.GlobusAPIError`

I thought it was a list of length 2,
- `ValueError`, a subclass of `ValueError`
- `globus_sdk.GlobusAPIError`

Can we scratch "a subclass of `ValueError`" from the list, but now on different grounds, which is that it's unnecessary to specify?

Comment (kurtmckee) [reply changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst:10]: One last alternate:

> `ValueError` (or a subclass) or `globus_sdk.GlobusAPIError`

?

**Review by sirosen — APPROVED** [review]

File: changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst
Comment (sirosen) [reply changelog.d/20241018_082221_kurtmckee_eliminate_none_return_values.rst:10]: My approval should speak for itself.

But since it doesn't seem to speak up, "yes!" and "w00t!" and other exclamations of joy. I now understand the text and it says everything you wanted it to say!

**Review by sirosen — APPROVED** [review]

**Pull request merged**
